ucp/_libs/ucx_api.pyx

```diff
@@ -446,11 +446,88 @@ cdef class UCXListener(UCXObject):
         return int(<uintptr_t>self._handle)
 
 
+# Counter used as UCXRequest UIDs
+cdef int _ucx_request_counter = 0
+
+cdef class UCXRequest:
+    """Python wrapper of UCX request handle.
+
+    Notice, this class doesn't own the handle and multiple instances of
+    UCXRequest can point to the same underlying UCX handle.
+    Furthermore, UCX can modify/free the UCX handle without notice
+    thus we use `_uid` to make sure the handle hasn't been modified.
+    """
+    cdef:
+        ucp_request *_handle
+        int _uid
+
+    def __init__(self, uintptr_t req_as_int):
+        global _ucx_request_counter
+        cdef ucp_request *req = <ucp_request*>req_as_int
+        assert req != NULL
+        self._handle = req
+
+        cdef dict info = {"status": "pending"}
+        if self._handle.info == NULL:  # First time we are wrapping this UCX request
+            Py_INCREF(info)
+            self._handle.info = <PyObject*> info
+            _ucx_request_counter += 1
+            self._uid = _ucx_request_counter
+            assert self._handle.uid == 0
+            self._handle.uid = _ucx_request_counter
+        else:
+            self._uid = self._handle.uid
```

Comment:
Could we add `weakref.finalize(self.close)` here somewhere to make sure this does get closed at some point?

Reply:
It is not easy to do; the problem is that multiple instances of `UCXRequest` can point to the same the underlying UCX request. I think the current implementation is reasonable safe since endpoints always makes sure to closes all `_inflight_msgs`.

Reply:
Though `self.close()` already protects against this, right?

Reply:
`self.close()` protects against double calls to `ucp_request_free()` not that we are calling `ucp_request_free` too early. 
Consider a `ucp_tag_send_nb()` call that **before returning** evoke the callback function with a UCX handle wrapped in `UCXRequest`. Now, this handles will goes out of scope and run the finalizer before  `ucp_tag_send_nb()` returns with the second handle.

I have considered using a *close counter* that could count to two but I think that is too much complexity for little gain.

Reply:
That makes sense. Thanks for the example. Appreciate this is not trivial.

Would [`axtexit.register`]( https://docs.python.org/3.8/library/atexit.html#atexit.register ) make sense or could things still be messed up by order of cleanup?

What happens if no one calls `close`?

Reply:
> Would [`axtexit.register`](https://docs.python.org/3.8/library/atexit.html#atexit.register) make sense or could things still be messed up by order of cleanup?

At exist, UCX might have freed the memory of the underlying UCX handler and result in a segfault. But remember, at exist all `UCXEndpoints` will cancel all pending `UCXRequest`.

> What happens if no one calls `close`?

Nothing too dangerous, UCX will print a warning saying that a `ucp_request` wasn't freed at exist.

Reply:
If it's already handled by some other cleanup mechanism, that sounds good enough to me.

Good to know. FWIW warnings are not totally free especially at scale as they can grind things to a halt. Leaking semaphores have bit us recently just as one example ( https://github.com/dask/distributed/issues/1078 )

That said, I'm ok leaving this as-is and just keeping an eye open for issues that may crop up.